[wip] feat(ui): Funding Pool v0

NunoAlexandre opened the pull request:
:construction: :building_construction: Under construction, please do not disturb :construction_worker_man: :hammer_and_wrench: 

Closes #974 
Superseds #952

NunoAlexandre commented on ui/src/funding/pool.ts:
```diff
@@ -76,7 +76,7 @@ export function make(wallet: Wallet): Pool {
     // only update members that have been added or removed
     const txs = [];
-    for (const address in settings.receiverAddresses) {
+    for (const address of settings.receiverAddresses) {
```
hacker :toolbox: :hammer: :wrench: :

CodeSandwich commented on FUNDING.md:
```diff
@@ -0,0 +1,90 @@
+- `npm run start` within `walletconnect-test-wallet`
+- `yarn start:ethereum` within `radicle-upstream`
+- `yarn start` within `radicle-upstream`
+
```
```suggestion

Remember to enable the funding feature in upstream configuration!
```

CodeSandwich reviewed (commented):
The funding docs are clear and straight forward :+1:

NunoAlexandre commented on FUNDING.md:
```diff
@@ -0,0 +1,86 @@
+- `yarn start` within `radicle-upstream`
+
+
+[wcw]: https://walletconnect.org/apps/
```
oops! thanks, Igor!

NunoAlexandre commented on FUNDING.md:
```diff
@@ -0,0 +1,86 @@
+  Ganachi provides an Ethereum RPC client for testing and development. We have
+  the Radicle Contracts deployed to and running on this instance. Here, we also
+  set an initial balance of the account we choose for development and debugging
+  purposes.
```
@CodeSandwich thank you for your input.

I agree that's a good thing to mention. I have added that bit at the end but kept the rest pretty much as it was. I didn't include the part relative to Infura as it's not relevant to use or set up the local ganache instance itself.

CodeSandwich replied:
Less overwhelming with information, I'm ok with that

CodeSandwich commented on FUNDING.md:
```diff
@@ -0,0 +1,94 @@
+
+- A local ganache instance
+
+  Ganachi provides a local Ethereum RPC client for testing and development. The
```
```suggestion
  Ganache provides a local Ethereum RPC client for testing and development. The
```

CodeSandwich commented on FUNDING.md:
```diff
@@ -0,0 +1,94 @@
+With everything installed and set up, run the following commands in different
+tabs:
+
+- `npm run start` within `walletconnect-test-wallet`
```
During the first ever run it'll be already running because of instructions in `Set up the local test ethereum account`
```suggestion
- `npm run start` within `walletconnect-test-wallet` if it's not already running
```

NunoAlexandre replied:
The instructions in the section you are referring to is to run `npm install`, which does not run the project. And I think this info is redundant, I guess everyone knows that if it's running they don't have to run it again

CodeSandwich replied:
I'm referring to this:
```
- Set up the local test ethereum account

  - Run `npm run start` within `walletconnect-test-wallet`. It should open the
    test wallet in your browser at `localhost:3000`.
```

I don't have a strong opinion on this, it's really a nitpick

geigerzaehler commented on ui/Hotkeys.svelte:
```diff
@@ -30,12 +30,7 @@
-      hasInputTarget ||
-      event.repeat
-    ) {
+    if (screen.isLocked() || hasInputTarget || event.repeat) {
```
Could you backport this and other unrelated changes to reduce the diff?

NunoAlexandre replied:
Done.

NunoAlexandre commented:
`Update`

At the moment, this branch contains all the functionality of the first MVP of the Radicle Funding Pools experience.

It also contains a few screens that regard the [attestationstep](https://github.com/radicle-dev/radicle-upstream/issues/965) step that will be put in place in the next iteration. Since it's all small and self-contained, I haven't dropped to avoid the duplicated work. Let me know if you don't feel cool about that.

Before setting this branch as ready for review:

- I am working out a way of testing this in a testnet and subsequently plan a release in the mainnet. Once we have the addresses of the contracts deployed in those lanes, we can mark this as done (after reviews, that is). 
- Alternatively to the previous point, we can cut it here and introduce the deployed contract addresses in a follow-up PR.
- @abbey-titcomb will review all the copyrighting

geigerzaehler reviewed (commented):
Here are some comments already

NunoAlexandre commented on ui/src/wallet.ts:
```diff
@@ -0,0 +1,259 @@
+  "http://localhost:8545"
+);
+
+export function build(): Wallet {
```
I think I have tried that before. If my memory serves, that didn't quite work because `class` cannot be serialized and stored in the svelte store. Will have a look cause indeed this is huge and super hard to follow. :+1:

NunoAlexandre replied:
Pretty much the same answer here as https://github.com/radicle-dev/radicle-upstream/pull/975#discussion_r555060510.

geigerzaehler replied:
I’m fine with it

NunoAlexandre commented on ui/src/transaction.ts:
```diff
@@ -0,0 +1,372 @@
+
+async function status(
+  receipt: TransactionReceipt
+): Promise<TxStatus | undefined> {
```
good catch!

NunoAlexandre commented on ui/src/transaction.ts:
```diff
@@ -0,0 +1,372 @@
+  });
+}
+
+async function status(
```
I got a good number of svelte errors on functions that returned `Promise` but weren't async. Will have a look

NunoAlexandre commented on ui/src/wallet.ts:
```diff
@@ -0,0 +1,259 @@
+  connect(): Promise<void>;
+  disconnect(): Promise<void>;
+  signer: ethers.Signer;
+}
```
yeah I will explore this :ok_hand:

NunoAlexandre replied:
I explored this but don't find the extra effort worth it in the end. Let me know if you feel strongly about it, I'd be happy to revisit this with you.

NunoAlexandre commented on ui/src/transaction.ts:
```diff
@@ -0,0 +1,372 @@
+  cap();
+}
+
+export function updateStatus(hash: string, status: TxStatus): void {
```
good catch, thanks. I have the idea these were shown as warnings before.

NunoAlexandre commented on ui/src/wallet.ts:
```diff
@@ -0,0 +1,259 @@
+      };
+      stateStore.set({ status: Status.Connected, connected });
+    } catch (error) {
+      console.error(error);
```
thanks. delete it, it was added for debugging purposes that are not generally needed.

NunoAlexandre commented on ui/Modal/Transaction.svelte:
```diff
@@ -0,0 +1,178 @@
+
+  let tx: Tx = undefined;
+
+  transactionsStore.subscribe(
+    txs => (tx = txs.find(x => x.hash === get(selectedStore)))
+  );
```
thanks for this catch. I have fixed this issue in other places but missed this one :hand:

NunoAlexandre commented on ui/src/transaction.ts:
```diff
@@ -0,0 +1,372 @@
+async function status(
+  receipt: TransactionReceipt
+): Promise<TxStatus | undefined> {
+  // TODO(nuno): Need to workout a way of detecting failed txs prior to the
+  // byzantium fork. Or might not be necessary at all.
```
I thought so but great to have your confirmation :partying_face:

rudolfs reviewed (commented):
Another magnum opus, can't wait to see this merged in! 👏 🚀 ⭐ 
Here are some thoughts, suggestions and questions.

### Visuals / feature work
- [ ] the funding enabled/disabled buttons only show the current selection when clicked twice

https://user-images.githubusercontent.com/158411/104049047-8c1c7280-51e4-11eb-8c6b-a4a2dc44dd81.mov

- [ ] the transaction center expand/collapse animation seems weird to me, was it designed like that?

https://user-images.githubusercontent.com/158411/104049725-b6bafb00-51e5-11eb-8d6f-98c0426bba0a.mov

- [ ] "plus" icon is not centered vertically:

<img width="479" alt="Screenshot 2021-01-08 at 13 57 31" src="https://user-images.githubusercontent.com/158411/104049779-d0f4d900-51e5-11eb-9a49-1dacda555dcb.png">

- [ ] there's no visual feedback between clicking on a transaction and it opening in a modal, sometimes it takes around a second or so to open


### Code
  - [ ] we should just use the `<Emoji>` component, instead of introducing new `<Illustration>` variants
  - [ ] could we avoid the underscore in imports? 
        `import * as _pool from "../../../../src/funding/pool";` -> `import * as fundingPool from "../../../../src/funding/pool";`
  - [ ] do we want to keep the debug logs?
```
src/wallet.ts
96:    console.log("Disconnect");

DesignSystem/Component/Funding/Pool/Receivers.svelte
23:    console.log("toggleReceiver", x);

DesignSystem/Component/Funding/Pool/Outgoing/Support.svelte
80:    console.log("onConfirmInWallet");
81:    console.log(`onConfirmInWallet, receivers ${[...receivers.entries()]}`);

DesignSystem/Component/TxButton.svelte
27:      console.log(JSON.stringify(error));
```

### Reminders about our conventions
  - [ ] we should use `dispatch` instead of passing down callbacks as props in components (`onCancel`, `onConfirm`, `onContinue`, `onBack`, `onConfirmed`, etc)
  - [ ] we should use `rem` instead of `px` for all new code
  - [ ] have only reusable components in the Design System and keep them dumb and composable, i.e. leave networking/data fetching to the parent

rudolfs commented:
Also noticed that we're now storing transactions in the electron local storage, I guess this is also important for our work on the multiple identities: https://github.com/radicle-dev/radicle-upstream/pull/1489.

And what's up with this prompt that defaults to agreeing to sending crash reports after a timeout when I now run `yarn`?

```
✔ Help us improve Hardhat with anonymous crash reports & basic usage data? (Y/n) · true
```

NunoAlexandre commented on ui/App.svelte:
```diff
@@ -101,6 +121,10 @@
 <NotificationFaucet />
 <Theme />
 
+{#if sessionIsUnsealed}
+  <TransactionCenter />
+{/if}
```
The transactions should be globally accessible, so as for the initial question regarding the place where they are "initialized", I find it fine there. The issue really is about z-indexing, and that is fair to discuss indeed. 

@brandonhaslegs how would you go about this? We could make a scale of z-indexes for each of the main screens and have them as global variables to avoid hardcoding them ad-hoc and lose the awareness of how one z-index relates to the rest of the system.

NunoAlexandre replied:
Fixed in https://github.com/radicle-dev/radicle-upstream/pull/975/commits/edc0ad58960ffca564cf79158fecdea10f0622a6

NunoAlexandre commented on ui/DesignSystem/Primitive/Icon/Link.svelte:
```diff
@@ -0,0 +1,12 @@
+<IconWrapper on:click {dataCy} {style}>
+  <path
+    d="M6.11043 8.99384C6.65933 8.93285 7.05486 8.43842 6.99387 7.88952C6.93288 7.34061 6.43846 6.94508 5.88955 7.00607L6.11043 8.99384ZM2.24998 9.74998L1.54287 9.04287H1.54287L2.24998 9.74998ZM7.0066 5.8852C6.9432 6.43383 7.33656 6.92998 7.8852 6.99338C8.43383 7.05678 8.92998 6.66341 8.99338 6.11478L7.0066 5.8852ZM15.75 2.24997L15.0429 2.95708V2.95708L15.75 2.24997ZM2.24999 15.75L2.95709 15.0429 ...
+    fill="#90A0AF" />
```
it's actually no longer used, deleted.

NunoAlexandre commented on ui/DesignSystem/Component/Funding/Pool/Receivers.svelte:
```diff
@@ -0,0 +1,132 @@
+  export let style = "";
+  export let alignment = "left"; // "center";
+
+  function toggleReceiver(x: pool.Address) {
```
done! can you tell I used to do Haskell? :laughing:

NunoAlexandre commented on ui/DesignSystem/Component/QR.svelte:
```diff
@@ -1,15 +1,16 @@
-  export let color = "var(--color-primary)";
+  export let key = "";
+  export let size = 225;
+  export let color = "var(--color-secundary)";
```
it should actually be black, thanks!

NunoAlexandre commented on ui/Modal/Funding/Onboarding.svelte:
```diff
@@ -0,0 +1,125 @@
+    );
+  }
+
+  /* Themz values */
```
This comment is part of the ES7 Nightly that parses the values below the comment and ensures that they are never null and that they are loved and cared for.

Jokes aside, that comment was just be typing out loud, sorry for the noise!

NunoAlexandre commented on ui/Screen/Profile/Funding.svelte:
```diff
@@ -0,0 +1,39 @@
+
+  // Hack to have Svelte working with checking the $wallet variant
+  // and thus be able to access its appropriate fields.
+  $: w = $wallet;
```
No, it's a hack so that svelte can realize the fine type it is handling. On line 27 we do:

``` svelte
{#if w.status === Status.Connected}
```

The scope within that condition has a `Connected` wallet at hand, but svelte is not able to figure that out and understand that `$wallet.connected.account` is an available field. It can't because we are referencing it as a store (`$wallet`). 

This hack basically just creates an alias to the store so that we can have Svelte figuring the fine type out while still working with the store. Another way to achieve this would be by casting the wallet as `any`, but that would blind us and easily lead to bugs and problems, so I figured this to be a better solution.

NunoAlexandre commented:
> And what's up with this prompt that defaults to agreeing to sending crash reports after a timeout when I now run `yarn`?
> 
> ```
> ✔ Help us improve Hardhat with anonymous crash reports & basic usage data? (Y/n) · true
> ```

wow what? I have never seen that and don't see it on my side when running `yarn`? :thinking:

rudolfs commented:
> wow what? I have never seen that and don't see it on my side when running yarn? 🤔

Try yanking your `node_modules` folder and then running `yarn`.

NunoAlexandre commented:
> > wow what? I have never seen that and don't see it on my side when running yarn? thinking
> 
> Try yanking your `node_modules` folder and then running `yarn`.

Gave it a run and still nothing. I wonder if it only happens on macOS. Do you see which dependency triggers that?

NunoAlexandre commented:
> the transaction center expand/collapse animation seems weird to me, was it designed like that?

I just picked the obsolete code we had before and neither @juliendonck nor @brandonhaslegs have commented on the way it is so I am checking that one.

NunoAlexandre commented:
> there's no visual feedback between clicking on a transaction and it opening in a modal, sometimes it takes around a second or so to open

solved in https://github.com/radicle-dev/radicle-upstream/pull/975#discussion_r553967639 :heavy_check_mark:

NunoAlexandre commented on ui/src/transaction.ts:
```diff
@@ -0,0 +1,372 @@
+    txs
+      .filter(tx => tx.status === TxStatus.AwaitingInclusion)
+      .forEach(async tx => {
+        const receipt = await provider.getTransactionReceipt(tx.hash);
```
Addressed in https://github.com/radicle-dev/radicle-upstream/pull/975/commits/ee4bb02c20d2318da78c10e6a227d1e159be82aa, we handle network failures :+1:

NunoAlexandre commented:
Thank you, @geigerzaehler and @rudolfs, for your reviews! 

I've addressed most of the issues pointed out. Here are the standing ones, for yours and mine clarity:

- [ ]  On the transaction center z-index https://github.com/radicle-dev/radicle-upstream/pull/975#discussion_r554897148 (@brandonhaslegs needed)
 - [ ] we should use `dispatch` instead of passing down callbacks as props in components (`onCancel`, `onConfirm`, `onContinue`, `onBack`, `onConfirmed`, etc)
 - [ ] we should use `rem` instead of `px` for all new code
 - [ ] On the issue with Hardhat granted permission issue (https://github.com/radicle-dev/radicle-upstream/pull/975#issuecomment-756929827) Still can't reproduce it, @rudolfs :thinking:

NunoAlexandre commented:
@rudolfs regarding:

>  - [ ] we should use `dispatch` instead of passing down callbacks as props in components (`onCancel`, `onConfirm`, `onContinue`, `onBack`, `onConfirmed`, etc)

I am not convinced that we should keep that "convention". By reading this [SO thread](https://stackoverflow.com/questions/61569655/svelte-event-forwarding-with-dispatcher-vs-passing-in-handling-function-which) I understand why dispatch is used (no other way for native components) but I find it a bad deal to, therefore, breach it out across the whole codebase for three main reasons:

1. it makes the code highly implicit and difficult to follow
2. the events are dynamically checked, resulting in the whole flow only being checked on runtime
3. no type-checking, one can call the wrong event or pass unfitting arguments and - again - only see it fail on runtime 

For these reasons, I suggest that we create a #housekeeping issue to migrate the usage of dispatch to pass functions as arguments instead, which resolves the 3 issues raised above.

Let me have your thoughts.

geigerzaehler commented:
On using callbacks over svelte events I agree with @NunoAlexandre. The latter only offers a short hand syntax for threading events through a stack of components while suffering from the shortcomings mentioned above. This makes callbacks a better option.

rudolfs commented:
@NunoAlexandre good points, especially about type-safety, didn't think of that. Also TIL that events allow for external inter-op, which we'll only need if we want to make our design system standalone (not a requirement for now).

Found [this proposal](https://github.com/sveltejs/svelte/issues/5602) in the Svelte issue tracker, it has a somewhat comprehensive list of the benefits and drawbacks of callbacks vs events .

It mentions:
  - [`on:` syntax](https://svelte.dev/docs#on_element_event) with modifiers like `preventDefault`, `stopPropagation`, `passive`, etc.
  - event forwarding
  - programmatic use
  - multiple listeners

as the benefits for events. I believe from those "features" we only use the `:on` syntax with modifiers and event forwarding in some places at the moment.

Found also this  [quote by one of Svelte's core devs](https://github.com/sveltejs/svelte/issues/2323#issuecomment-477184095): "It's now a lot easier to have props that are callbacks in Svelte 3, so you'll probably be using events less anyway."

So I guess this means disregard my initial suggestion.
Looking forward to removing some implicitness and runtime exceptions.

rudolfs commented:
> On the issue with Hardhat granted permission issue (#975 (comment)) Still can't reproduce it, @rudolfs 🤔

This is odd, I just did a `rm -rf node_modules && yarn` again on my Intel macOS 10.15.7 machine and it happened again:

<details>
<summary>
terminal output
</summary>

```
yarn install v1.22.5
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[3/4] 🔗  Linking dependencies...
[4/4] 🔨  Building fresh packages...
$ yarn prepack
yarn run v1.22.5
$ yarn build
$ hardhat compile && tsc
✖ Help us improve Hardhat with anonymous crash reports & basic usage data? (Y/n) · true
(node:70526) Warning: Accessing non-existent property 'INVALID_ALT_NUMBER' of module exports inside circular dependency
(Use `node --trace-warnings ...` to show where the warning was created)
(node:70526) Warning: Accessing non-existent property 'INVALID_ALT_NUMBER' of module exports inside circular dependency
Downloading compiler 0.7.6
Compiling 35 files with 0.7.6
@ensdomains/ens/contracts/ENS.sol: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.

@uniswap/lib/contracts/libraries/FixedPoint.sol: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.

@uniswap/v2-core/contracts/interfaces/IUniswapV2Factory.sol: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.

@uniswap/v2-core/contracts/interfaces/IUniswapV2Pair.sol: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.

contracts/DummyRouter.sol: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.

contracts/libraries/SafeMath.sol: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.

contracts/libraries/Uniswap.sol: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.

@ensdomains/ens/contracts/ENS.sol:20:38: Warning: This declaration has the same name as another declaration.
    function setRecord(bytes32 node, address owner, address resolver, uint64 ttl) external;
                                     ^-----------^
@ensdomains/ens/contracts/ENS.sol:27:5: The other declaration is here:
    function owner(bytes32 node) external view returns (address);
    ^-----------------------------------------------------------^

@ensdomains/ens/contracts/ENS.sol:20:53: Warning: This declaration has the same name as another declaration.
    function setRecord(bytes32 node, address owner, address resolver, uint64 ttl) external;
                                                    ^--------------^
@ensdomains/ens/contracts/ENS.sol:28:5: The other declaration is here:
    function resolver(bytes32 node) external view returns (address);
    ^--------------------------------------------------------------^

@ensdomains/ens/contracts/ENS.sol:20:71: Warning: This declaration has the same name as another declaration.
    function setRecord(bytes32 node, address owner, address resolver, uint64 ttl) external;
                                                                      ^--------^
@ensdomains/ens/contracts/ENS.sol:29:5: The other declaration is here:
    function ttl(bytes32 node) external view returns (uint64);
    ^--------------------------------------------------------^

@ensdomains/ens/contracts/ENS.sol:21:60: Warning: This declaration has the same name as another declaration.
    function setSubnodeRecord(bytes32 node, bytes32 label, address owner, address resolver, uint64 ttl) external;
                                                           ^-----------^
@ensdomains/ens/contracts/ENS.sol:27:5: The other declaration is here:
    function owner(bytes32 node) external view returns (address);
    ^-----------------------------------------------------------^

@ensdomains/ens/contracts/ENS.sol:21:75: Warning: This declaration has the same name as another declaration.
    function setSubnodeRecord(bytes32 node, bytes32 label, address owner, address resolver, uint64 ttl) external;
                                                                          ^--------------^
@ensdomains/ens/contracts/ENS.sol:28:5: The other declaration is here:
    function resolver(bytes32 node) external view returns (address);
    ^--------------------------------------------------------------^

@ensdomains/ens/contracts/ENS.sol:21:93: Warning: This declaration has the same name as another declaration.
    function setSubnodeRecord(bytes32 node, bytes32 label, address owner, address resolver, uint64 ttl) external;
                                                                                            ^--------^
@ensdomains/ens/contracts/ENS.sol:29:5: The other declaration is here:
    function ttl(bytes32 node) external view returns (uint64);
    ^--------------------------------------------------------^

@ensdomains/ens/contracts/ENS.sol:22:59: Warning: This declaration has the same name as another declaration.
    function setSubnodeOwner(bytes32 node, bytes32 label, address owner) external returns(bytes32);
                                                          ^-----------^
@ensdomains/ens/contracts/ENS.sol:27:5: The other declaration is here:
    function owner(bytes32 node) external view returns (address);
    ^-----------------------------------------------------------^

@ensdomains/ens/contracts/ENS.sol:23:40: Warning: This declaration has the same name as another declaration.
    function setResolver(bytes32 node, address resolver) external;
                                       ^--------------^
@ensdomains/ens/contracts/ENS.sol:28:5: The other declaration is here:
    function resolver(bytes32 node) external view returns (address);
    ^--------------------------------------------------------------^

@ensdomains/ens/contracts/ENS.sol:24:37: Warning: This declaration has the same name as another declaration.
    function setOwner(bytes32 node, address owner) external;
                                    ^-----------^
@ensdomains/ens/contracts/ENS.sol:27:5: The other declaration is here:
    function owner(bytes32 node) external view returns (address);
    ^-----------------------------------------------------------^

@ensdomains/ens/contracts/ENS.sol:25:35: Warning: This declaration has the same name as another declaration.
    function setTTL(bytes32 node, uint64 ttl) external;
                                  ^--------^
@ensdomains/ens/contracts/ENS.sol:29:5: The other declaration is here:
    function ttl(bytes32 node) external view returns (uint64);
    ^--------------------------------------------------------^

@ensdomains/ens/contracts/ENS.sol:31:31: Warning: This declaration has the same name as another declaration.
    function isApprovedForAll(address owner, address operator) external view returns (bool);
                              ^-----------^
@ensdomains/ens/contracts/ENS.sol:27:5: The other declaration is here:
    function owner(bytes32 node) external view returns (address);
    ^-----------------------------------------------------------^

@openzeppelin/contracts/token/ERC20/ERC20.sol:55:5: Warning: Visibility for constructor is ignored. If you want the contract to be non-deployable, making it "abstract" is sufficient.
    constructor (string memory name_, string memory symbol_) public {
    ^ (Relevant source part starts here and spans across multiple lines).

contracts/DummyUpgradable.sol: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.

contracts/ErrorReporter.sol: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.

contracts/ProxyAdminStorage.sol: Warning: SPDX license identifier not provided in source file. Before publishing, consider adding a comment containing "SPDX-License-Identifier: <SPDX-License>" to each source file. Use "SPDX-License-Identifier: UNLICENSED" for non-open-source code. Please see https://spdx.org for more information.

Compilation finished successfully
Successfully generated Typechain artifacts!
✨  Done in 36.65s.
[3/4] 🔗  Linking dependencies...
warning " > rollup-plugin-node-externals@2.2.0" has unmet peer dependency "builtin-modules@^3.1.0".
[4/4] 🔨  Building fresh packages...
[1/10] ⠈ secp256k1
[-/10] ⠈ waiting...
[-/10] ⠈ waiting...
[-/10] ⠈ waiting...
^C/10] ⠈ waiting...
```
</details>

And when I tried it on my macOS 11.1 machine it didn't happen. Then I tried to make a screen recording of the machine where it appears and it didn't show up anymore. 🤷

I wonder if it was something with my global yarn cache?

geigerzaehler commented:
> On the issue with Hardhat granted permission issue (#975 (comment)) Still can't reproduce it, @rudolfs thinking

I’ve discovered this file: `~/.local/share/hardhat-nodejs/analytics.json`.

rudolfs commented:
Yeah, I've got this on my mac as well in:
```
./Library/Application Support/hardhat-nodejs
./Library/Application Support/hardhat-nodejs/analytics.json
./Library/Preferences/hardhat-nodejs
```

NunoAlexandre commented:
@rudolfs @geigerzaehler thanks for looking into the hardhat issue on your machines. We don't have that package in `Cargo(.lock)` neither in `node_modules`. So I presume this is indeed some global dependency you guys have that I don't have and that shouldn't impact the project in any way. Checking that one out

rudolfs commented:
@NunoAlexandre thanks for looking into it. I think I found the origin, it's a transitive dependency and it comes from `radicle-contracts`: https://github.com/radicle-dev/radicle-contracts/blob/master/package.json#L24.

Also found a way to reliably replicate getting this confirmation dialog. I have to delete all caches via:
```
rm -rf ~/Library/Application Support/hardhat-nodejs
rm -rf ~/Library/Preferences/hardhat-nodejs
rm -rf ~/Library/Caches/hardhat-nodejs

yarn cache clean
rm -rf ~/Library/Caches/Yarn/v6

cd radicle-upstream
rm -rf node_modules
yarn

yarn install v1.22.5
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
[1/4] 🔍  Resolving packages...
[2/4] 🚚  Fetching packages...
error https://registry.yarnpkg.com/rxjs/-/rxjs-6.6.3.tgz: Extracting tar content of undefined failed, the file appears to be corrupt: "ENOENT: no such file or directory, chmod '/Users/rudolfs/Library/Caches/Yarn/v6/npm-rxjs-6.6.3-8ca84635c4daa900c0d3967a6ee7ac60271ee552-integrity/node_modules/rxjs/_esm5/internal/operators/window.js.map'"
info Visit https://yarnpkg.com/en/docs/cli/install for documentation about this command.
[3/4] 🔗  Linking dependencies...
[4/4] 🔨  Building fresh packages...
$ yarn prepack
yarn run v1.22.5
$ yarn build
$ hardhat compile && tsc
✖ Help us improve Hardhat with anonymous crash reports & basic usage data? (Y/n) · true
```

I guess we can also dismiss this, I just wanted to make sure this is a legit dependency and not some malware we're pulling in.

rudolfs commented:
Maybe we can opt-out of the analytics via some config `radicle-contracts`? \cc @cloudhead @CodeSandwich 

https://github.com/nomiclabs/hardhat/issues/761

NunoAlexandre commented:
@geigerzaehler @rudolfs thanks a ton for your reviews! Everything is now addressed :tada: Please, have another look!

rudolfs commented on ui/DesignSystem/Primitive/Icon/Plus.svelte:
```diff
@@ -3,9 +3,11 @@
   export let dataCy = "";
   export let style = "";
+
+  $: iconStyle = `${style  }; display: flex;`;
```
This should probably just go in the `IconWrapper` component itself.

rudolfs reviewed (commented):
Getting close.

- [ ] we should stick to our standard way of showing error notifications (`error.ts show()`) so they include a copyable error message

Error notifications in funding:
<img width="1024" alt="Screenshot 2021-01-15 at 10 18 11" src="https://user-images.githubusercontent.com/158411/104706248-23b91e00-571b-11eb-97d3-f50f5c9c861c.png">

vs

Errors in coco:

<img width="960" alt="Screenshot 2021-01-15 at 10 18 58" src="https://user-images.githubusercontent.com/158411/104706277-2c115900-571b-11eb-8779-89e847e0b4d2.png">


```
{
  "code": "ProjectCreationFailure",
  "message": "Could not create project: the path provided '/Users/rudolfs/work/radicle-tests/new' already exists",
  "source": {
    "response": {}
  }
}
```
- [ ] support form resets when two edit actions are performed

https://user-images.githubusercontent.com/158411/104706638-a80ba100-571b-11eb-9bdd-3fd945af1bf7.mov

rudolfs commented on ui/DesignSystem/Primitive/Icon/Plus.svelte:
```diff
@@ -3,9 +3,11 @@
   export let dataCy = "";
   export let style = "";
+
+  $: iconStyle = `${style}; display: flex;`;
```
This should probably just go in the IconWrapper component itself.

rudolfs commented:
One question regarding the release of the feature: if it can only be used in development mode because of the requirements of running it with `ethereum:start`, local walletconnect wallet and the `sandbox/.local-eth-account`, what's the point of having the feature flag in the settings if this can't be used in a packaged production app?

Or will there be follow-up PRs to address this before we release it to the public?

rudolfs commented on FUNDING.md:
```diff
@@ -0,0 +1,95 @@
+
+  - Finally, paste the copied address in that file
+
+- Install `ganache-cli`:
+
+  `npm install -g ganache-cli`
```
Do we really need this?
We already have it in our `devDependencies`.

geigerzaehler commented on ui/src/funding/contract.ts:
```diff
@@ -0,0 +1,92 @@
+export class PoolContract {
+  contract: Erc20Pool;
+
+  constructor(signer: Signer, address: string = POOL_ADDRESS) {
```
It seems to me that we never call this constructor with just one argument. The `pool` function handles the default value. This means we can just remove the default.

geigerzaehler commented on ui/src/funding/contract.ts:
```diff
@@ -0,0 +1,92 @@
+
+export function pool(
+  signer: Signer,
+  address: string = POOL_ADDRESS
```
It appears that this function is never called with a second argument. Can we remove it?

geigerzaehler commented on ui/src/funding/pool.ts:
```diff
@@ -0,0 +1,335 @@
+  Present = "Present",
+}
+
+export function make(wallet: Wallet): Pool {
```
Is there are reason why `Pool` is not a class? This might make things more readable since method definition is closer to method implementation.

NunoAlexandre replied:
For no reason in particular, besides my trauma from classes, I guess. I'd prefer to have this as a follow-up piece of course if that's alright with you :)

geigerzaehler commented on ui/src/funding/pool.ts:
```diff
@@ -0,0 +1,335 @@
+export interface Pool {
+  data: remote.Store<PoolData>;
+
+  // Get the account that owns this pool. Should be the connected wallet account.
```
> Should be the connected wallet account.
	
What does this mean? How does this influence the return value or potential exceptions?

NunoAlexandre replied:
The comment itself does not influence the return value, it just informs the reader that the account owning the pool instance we have at hand is (that is, "should be") the connected wallet account. So doing `pool.getAccount` should be the same as doing `$wallet.connected.account` (when the wallet is connected, of course).

Let me know how this can be clearer :+1:

NunoAlexandre replied:
I've cut the second sentence out, let me have your thoughts.

geigerzaehler replied:
Fine with me.

Just to clarify where the confusion came from: It was not clear what the subject of the sentence was and who would be responsible for ensuring the requirement and how it would influence anything regarding the method.

geigerzaehler commented on ui/src/funding/pool.ts:
```diff
@@ -0,0 +1,335 @@
+  Present = "Present",
+}
+
+export function make(wallet: Wallet): Pool {
```
:+1: on using an explicit `wallet` dependency rather than importing it from the wallet module.

NunoAlexandre commented on ui/DesignSystem/Component/SidebarLayout.svelte:
```diff
@@ -15,9 +15,8 @@
     position: relative;
     left: var(--sidebar-width);
-    width: calc(100vw - var(--sidebar-width));
+    min-width: calc(100vw - var(--sidebar-width));
     height: 100%;
-    min-width: 60rem;
```
As described in the commit introducing this change, this was something discussed with Brandon. The app is not designed nor engineered, atm, to be responsive, and this change ensures that 1) the layout by definition always takes the whole viewport and 2) that if the content of the screen requires more space, that it should become scrollable.

I am not sure why - considering the size of your viewport - it is becoming scrollable for you since it has enough space to not go that route, I am not seeing that myself but I will have a look asaic.

rudolfs replied:
> I am not sure why - considering the size of your viewport - it is becoming scrollable for you since it has enough space to not go that route, I am not seeing that myself but I will have a look asaic.

I only raised this because there's enough space on my viewport (15" mbp) to display everything without needing a horizontal scrollbar. So you'll probably have to look into solution 2.

brandonhaslegs replied:
>  Screen.Recording.2021-01-15.at.09.52.15.mov

It's not clear to me why this is happening. The support panels just have a min-width that is too large right? The min width doesn't need to be this big, as you can see, none of the UI is colliding with itself yet.

It appears to me we have this much space (see red arrow) left to continue condensing the UI. So why does the min-width stop here?
![image](https://user-images.githubusercontent.com/4406983/104898509-059d2900-597a-11eb-8e8b-c6838c5d6946.png)

NunoAlexandre replied:
On my machine, I only get the scrolling when the components no longer fit inline, so I am not sure either why this is happening. @rudolfs (on mac) realized that if we resize the window back and forth that the issue disappears. I'm having a look.

NunoAlexandre replied:
Fixed in https://github.com/radicle-dev/radicle-upstream/pull/975/commits/e3d30854855b53ddfdcd1c120a786389d46123ec

geigerzaehler commented on ui/src/funding/pool.ts:
```diff
@@ -0,0 +1,335 @@
+export function isOnboarded(data: PoolData): boolean {
+  return (
+    data.erc20Allowance > 0 &&
+    (data.receivers.size > 0 || data.amountPerBlock !== "0" || data.balance > 0)
```
I’m not sure the two last comparisons work as intended because the left-hand side is `BigNumberish`. For example if it is a `BigNumber` that represents zero than the last comparison returns true (I think).

To address this I’d suggest making `amountPerBlock` and `balance` of type `BigNumber` and use the appropriate methods. In fact, everything that is of type `BigNumberish` should probably be of type `BigNumber`. This would reduce the number of cases that would need to be distinguished and would make it easier to handle such values.

geigerzaehler commented on ui/src/funding/pool.ts:
```diff
@@ -0,0 +1,335 @@
+// Convert `Receivers` to `ReceiverWeight[]`, the latter being the
+// representation receivers have in the Radicle Contracts.
+function toReceiverWeights(receivers: Receivers): contract.PoolReceiver[] {
+  return [...receivers.entries()].map(([address, status]) => {
```
Is `.entries()` necessary? Isn’t `[...receivers]` enough?

geigerzaehler commented on ui/src/validation.ts:
```diff
@@ -44,7 +44,7 @@ interface FirstHandleCharOptions {
 }
 
-interface FormatConstraints {
+export interface FormatConstraints {
```
This change is not necessary.

geigerzaehler commented on ui/src/funding/pool.ts:
```diff
@@ -0,0 +1,335 @@
+        transaction.add(
+          transaction.updateSupport(tx, amountPerBlock, newReceivers)
+        );
+        tx.wait();
```
Aren’t we missing an `await` or `return` here? Otherwise `finally` will be called _before_ the transaction has been submitted.

NunoAlexandre replied:
Where exactly? I don't think that's the case that `finally` is called before submission but I will check!

NunoAlexandre replied:
Verified that `finally` only runs when the transaction is either approved or rejected in the wallet.

geigerzaehler replied:
> Verified that `finally` only runs when the transaction is either approved or rejected in the wallet.

That’s correct. The promise returned from `.updatePlan()` only resolves when the transaction has been submitted to the chain successfully. But why do we call `tx.wait()` here and don’t wait for it?

NunoAlexandre replied:
Ah, I now get your question :) I believe this is out you have set this up initially, so I just followed the lead here. I am not sure if we need to `.wait` at all since we will be looking up the statuses of ongoing transactions ya. What do you recon?

geigerzaehler replied:
> I am not sure if we need to `.wait` at all since we will be looking up the statuses of ongoing transactions ye.

I don’t think we need to `.wait()` either. Let’s just remove this line.

NunoAlexandre replied:
Done!

NunoAlexandre commented on ui/DesignSystem/Component/Illustration.svelte:
```diff
@@ -12,43 +12,44 @@
   }
```
These changes are generally useful, as this component now is programmed to only support the two Illustrations we use at the moment. So the changes are an improvement but if you prefer to reduce the number of changes with the risk of having to re-introduce them later I'm ok with that.

rudolfs replied:
Alright, then let's leave it as is.

geigerzaehler commented on ui/src/funding/pool.ts:
```diff
@@ -0,0 +1,335 @@
+        contract.POOL_ADDRESS
+      );
+    } else {
+      return Promise.resolve(0);
```
```suggestion
      return 0;
```

NunoAlexandre commented:
> * we should stick to our standard way of showing error notifications (`error.ts show()`) so they include a copyable error message

Thanks for seeing this! Unlike coco, we can't so trivially have structured errors in this case but I will see what we can do :)

NunoAlexandre commented:
> One question regarding the release of the feature: if it can only be used in development mode because of the requirements of running it with `ethereum:start`, local walletconnect wallet and the `sandbox/.local-eth-account`, what's the point of having the feature flag in the settings if this can't be used in a packaged production app?
> 
> Or will there be follow-up PRs to address this before we release it to the public?

I am with you on this. The idea behind the feature flag was to have the feature available to end users but only for a selected or curious few who would like to give it a spin and help us validate it. For that, indeed, we'd need to have it pointing to a testnet.

So, answering your question, there will be a follow-up PR offering an environment option for the funding feature (something like `localhost` / `testnet`) before it should be made public. However, I don't want to have this becoming a blocker to release other app deliverables, so we could, for now, have it behind the `isExperimental` flag until said PR comes in.

How does that sound?

NunoAlexandre commented:
@rudolfs @geigerzaehler thanks for the second round! I have addressed/replied to all your comments.

Waiting on @brandonhaslegs to have a look at https://github.com/radicle-dev/radicle-upstream/pull/975#discussion_r557536606.

rudolfs commented on package.json:
```diff
@@ -65,8 +65,11 @@
+    "@rollup/plugin-json": "^4.1.0",
     "@rollup/plugin-commonjs": "^17.0.0",
-    "@rollup/plugin-node-resolve": "^11.0.1",
+    "@rollup/plugin-node-resolve": "^10.0.0",
```
Is there a reason we're rolling back, or is this just a merge snafu?
We recently upgraded all packages in master, might be a good idea to run:
```
yarn upgrade-interactive --latest
yarn upgrade
```
just to make sure we're still on the latest & greatest.

NunoAlexandre replied:
Yeah, the latest version does not go well with some dependency we are using. trying to remember which one 🤔

rudolfs replied:
Might be a good idea to document this somehow, so we don't accidentally upgrade it and break stuff.

rudolfs replied:
Or perhaps we just pin it to the specific version, so upgrades don't touch it?

NunoAlexandre replied:
If we bump it to `11.x.x.`, we will see a clear error message when running the app. For me, that'd be enough and I assume CI would fail too. Let me know!

rudolfs commented on FUNDING.md:
```diff
@@ -0,0 +1,86 @@
+
+- `npm run start` within `walletconnect-test-wallet`
+- `yarn ethereum:start` within `radicle-upstream`
+- `yarn start` within `radicle-upstream`
```
```suggestion
`RADICLE_UPSTREAM_EXPERIMENTAL=true yarn start` within `radicle-upstream`
```

rudolfs reviewed (commented):
Great work!
I think once we sort out https://github.com/radicle-dev/radicle-upstream/pull/975#discussion_r557536606, it's a "go" from my side.

rudolfs reviewed (dismissed):
LGTM 🚀

geigerzaehler reviewed (dismissed):
:money_with_wings:

MeBrei commented on ui/DesignSystem/Primitive/Icon/CurrencyDAI.svelte:
```diff
@@ -0,0 +1,18 @@
+  <path
+    fill-rule="evenodd"
+    clip-rule="evenodd"
+    d=" M 12.981 6.5 C 12.994 6.665 13 6.832 13 7 L 13 8 C 13 8.168 12.994 8.335 12.981 8.5 L 14 8.5 C 14.552 8.5 15 8.948 15 9.5 C 15 10.052 14.552 10.5 14 10.5 L 12.502 10.5 C 11.522 12.849 9.204 14.5 6.5 14.5 L 4.5 14.5 C 3.395 14.5 2.5 13.604 2.5 12.5 L 2.5 12.23 C 2.5 11.678 2.948 11.23 3.5 11.23 C 4.052 11.23 4.5 11.678 4.5 12.23 L 4.5 12.5 L 6.5 12.5 C 8.06 12.5 9.435 11.706 10.242 10.5 L  ...
```
Currently our pre commit hook does not break lines inside too long strings, so it would need to be split manually.

MeBrei commented on ui/Modal/Funding/Link/EthToRadicle.svelte:
```diff
@@ -0,0 +1,106 @@
+  }
+</style>
+
+<div class="wrapper">
+  <Emoji emoji="🧦" size="huge" />
+
+  <header>
+    <h1 style="margin-top: 1.5rem;">Add your Ethereum account to Radicle</h1>
+    <p
+      style="margin-top: 1.5rem; padding: 0 var(--content-padding)"
+      class="typo-text">
+      Are you sure you want to add your Ethereum address to your Radicle
+      account?
+    </p>
+  </header>
```
Could this be extracted as it is the same for `EnterPassphrase` and `RadicleToEth`? 
This would make editing eg the text confined to one space

NunoAlexandre replied:
Good catch!

NunoAlexandre replied:
@MeBrei actually, we are unsure whether `EnterPassword` will be used in the end so I'd rather leave it as is and wait another few weeks to get back to this work and improve what we actually end up with.

MeBrei commented on ui/Screen/Funding/Pool/Outgoing/GetStarted.svelte:
```diff
@@ -0,0 +1,43 @@
+
+<div class="wrapper">
+  <Emoji emoji="💸" size="huge" />
+  <p class="typo-text">Stream digital money your favorite people.</p>
```
Is there a `to` missing in the sentence? I saw that it is the same in the figma screen.
```suggestion
  <p class="typo-text">Stream digital money to your favorite people.</p>
```

NunoAlexandre commented:
@MeBrei @rudolfs @geigerzaehler appreciate your greenest lights and get this baby monster merged :smiling_face_with_three_hearts:
